logtail/logtail.go

```diff
@@ -573,169 +596,211 @@ func (l *Logger) sendLocked(jsonBlob []byte) (int, error) {
+			dst = append(dst, ',')
 		}
-		b = bytes.TrimRight(b, ",")
-		b = append(b, "}, "...)
+		if errDetail != "" || errData != nil {
+			dst = append(dst, `"error":{`...)
+			if errDetail != "" {
+				dst = append(dst, `"detail":`...)
+				dst, _ = jsontext.AppendQuote(dst, errDetail)
+				dst = append(dst, ',')
+			}
+			if errData != nil {
+				dst = append(dst, `"bad_data":`...)
+				dst = append(dst, errData...)
+				dst = append(dst, ',')
+			}
+			dst = bytes.TrimRight(dst, ",")
+			dst = append(dst, '}', ',')
+		}
+		dst = bytes.TrimRight(dst, ",")
+		dst = append(dst, '}', ',')
 	}
 
-	if l.metricsDelta != nil {
+	// Append optional metrics metadata.
+	if !skipMetrics && l.metricsDelta != nil {
 		if d := l.metricsDelta(); d != "" {
-			b = append(b, `"metrics": "`...)
-			b = append(b, d...)
-			b = append(b, `",`...)
+			dst = append(dst, `"metrics":"`...)
+			dst = append(dst, d...)
+			dst = append(dst, '"', ',')
 		}
 	}
 
-	// Add the log level, if non-zero. Note that we only use log
-	// levels 1 and 2 currently. It's unlikely we'll ever make it
-	// past 9.
+	// Add the optional log level, if non-zero.
+	// Note that we only use log levels 1 and 2 currently.
+	// It's unlikely we'll ever make it past 9.
 	if level > 0 && level < 10 {
-		b = append(b, `"v":`...)
-		b = append(b, '0'+byte(level))
-		b = append(b, ',')
+		dst = append(dst, `"v":`...)
+		dst = append(dst, '0'+byte(level))
+		dst = append(dst, ',')
 	}
-	b = append(b, "\"text\": \""...)
-	for _, c := range buf {
-		switch c {
-		case '\b':
-			b = append(b, '\\', 'b')
-		case '\f':
-			b = append(b, '\\', 'f')
-		case '\n':
-			b = append(b, '\\', 'n')
-		case '\r':
-			b = append(b, '\\', 'r')
-		case '\t':
-			b = append(b, '\\', 't')
-		case '"':
-			b = append(b, '\\', '"')
-		case '\\':
-			b = append(b, '\\', '\\')
-		default:
-			// TODO: what about binary gibberish or non UTF-8?
-			b = append(b, c)
-		}
+
+	return dst
+}
+
+// appendText appends a raw text message in the Tailscale JSON log entry format.
+func (l *Logger) appendText(dst, src []byte, skipClientTime bool, procID uint32, procSequence uint64, level int) []byte {
+	dst = slices.Grow(dst, len(src))
+	dst = append(dst, '{')
+	dst = l.appendMetadata(dst, skipClientTime, false, procID, procSequence, "", nil, level)
+	if len(src) == 0 {
+		dst = bytes.TrimRight(dst, ",")
+		return append(dst, "}\n"...)
 	}
-	if nTruncated > 0 {
-		b = append(b, "…+"...)
-		b = strconv.AppendInt(b, int64(nTruncated), 10)
+
+	// Append the text string, which may be truncated.
+	// Invalid UTF-8 will be mangled with the Unicode replacement character.
+	max := 16 << 10
+	if l.lowMem {
+		max = 4 << 10
 	}
-	b = append(b, "\"}\n"...)
-	return b
+	dst = append(dst, `"text":`...)
+	dst = appendTruncatedString(dst, src, max)
+	return append(dst, "}\n"...)
+}
+
+// appendTruncatedString appends a JSON string for src,
+// truncating the src to be no larger than n.
+func appendTruncatedString(dst, src []byte, n int) []byte {
```

Comment:
This method has a strange feel, if dst and n aren't carefully pre-arranged it might allocate in the long string case, as the truncation doesn't account for the 6+ bytes of overhead on the tail. Do we want to estimate the tail size up front?

Reply:
We could, but correctly enforcing max lengths after mutation is tricky to get right, and we already fail to do this level of strict guarantee in other places in this package.

It seems that this package has historically aimed for a soft adherence to the 256KiB max limit, with the assumption that the true limit on the log server side is actually 512KiB. So long as our munging never doubles the size, then we're okay.

Reply:
Also, even within this function itself, making the limit based on the output would need to take into consideration the effects of escaping certain runes. It can get fairly complicated for little gain.